### PR TITLE
Fix results links are active if regular user is not participant in contest

### PR DESCRIPTION
**Pull request by shsimeonova**
Closes https://github.com/SoftUni-Internal/exam-systems-issues/issues/1343

### CODE REVIEW

File: Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs [hunk 1]
```diff
@@ -91,34 +91,46 @@ public async Task<ContestDetailsServiceModel> GetContestDetails(int id)
...
 
-            contestDetailsServiceModel!.CanBeCompeted = contestActivityEntity.CanBeCompeted;
-            contestDetailsServiceModel.CanBePracticed = contestActivityEntity.CanBePracticed;
-            contestDetailsServiceModel.IsAdminOrLecturerInContest = isLecturerOrAdmin;
+            var participantToGetProblemsFrom = contestDetailsServiceModel.CanBeCompeted
+                ? competeParticipant
+                : await this.participantsData.GetWithProblemsForParticipantsByContestByUserAndIsOfficial(
+                    id,
+                    user.Id,
+                    false);
```
Comment (gogo4ds) [inline Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs:109]: Why did you split this into 2 variables? In the end is always a single participant that is used. Just use `contestActivityEntity.CanBeCompeted` for the variable. Also the same boolena can be used where you check `competeParticipant == null`. If the contest can be competed by the participant, this means compete participant must be present, or I'm missing something.

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs (hunk 1 above)
Comment (shsimeonova) [reply Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs:109]: In the details page the problems that are shown are based on the participant. I split the participant variable so i can determine if the user can view compete or practice results separately.

Comment (gogo4ds) [reply Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs:109]: So why not use the activityService.CanBeCompeted for this?

File: Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs [hunk 2]
```diff
@@ -91,34 +91,43 @@ public async Task<ContestDetailsServiceModel> GetContestDetails(int id)
...
+            var userParticipants = await this.participantsData
+                .GetWithProblemsForParticipantsByContestByUser(id, user.Id)
+                .ToListAsync();
 
-            if (!isLecturerOrAdmin && participant != null && contestActivityEntity.CanBeCompeted)
+            var competeParticipant = userParticipants.FirstOrDefault(p => p.IsOfficial);
+
+            var participantToGetProblemsFrom = contestDetailsServiceModel.CanBeCompeted
+                ? competeParticipant
+                : userParticipants.FirstOrDefault(p => !p.IsOfficial);
```
Comment (gogo4ds) [inline Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs:106]: Can you extract `practiceParticiapnt` as well..
Also it's better to use `contestActivityEntity.CanBeCompeted` and set the `contestDetailsServiceModel` properties at the end of the method. This way you minimize the possibility for someone misusing some of the properties before being set and also it's more clear that the `contestActivityEntity.CanBeCompeted` is taking into account the user itself.

File: Services/UI/OJS.Services.Ui.Data/IParticipantsDataService.cs [hunk 3]
```diff
@@ -14,6 +14,8 @@ public interface IParticipantsDataService : IDataService<Participant>
 
         Task<Participant?> GetWithContestAndSubmissionDetailsByContestByUserAndIsOfficial(int contestId, string userId, bool isOfficial);
 
+        IQueryable<Participant> GetWithProblemsForParticipantsByContestByUser(int contestId, string userI);
```
Comment (gogo4ds) [inline Services/UI/OJS.Services.Ui.Data/IParticipantsDataService.cs:17]: `..ByContestAndUser`

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs (hunk 2 above)
Comment (shsimeonova) [reply Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs]: Fixed

**Review by gogo4ds — APPROVED** [review]

**Pull request merged**